src/utils/adjust-markup.js

```diff
@@ -0,0 +1,144 @@
+const PLACEHOLDER_TEMPLATE = /\$\{\d+\}/g;
+
+/**
+ * Loads the markup config into the action, if present.
+ *
+ * @param {import("../context").Context} context some param
+ * @param {import("../context").Action} action some other param
+ */
+async function loadConfig(context, action) {
+  const config = new MarkupConfig();
+  const hasFile = await config.hasFile();
+  action.markupconfig = action.markupconfig || (hasFile && config.init()) || undefined;
+}
+
+/**
+ * Returns the HTML element for the provided HTML template.
+ *
+ * @param {String} template The HTML template to use
+ *
+ * @returns {HTMLElement} the resulting HTML element including a `${0}` placeholder
+ */
+function getHTMLElement(template) {
+  const html = expand(template);
+  const doc = new JSDOM().window.document;
+  const dom = doc.createElement('div');
+  dom.innerHTML = html;
+  return dom.firstChild;
+}
+
+/**
+ * Adjust the MDAST tree according to the markup config.
+ * This is done by registering new matchers on the VDOMTransformer before the HTML
+ * is generated in the pipeline
+ *
+ * @param {Object} context the execution context
+ * @param {Object} logger the pipeline logger
+ * @param {Object} transformer the VDOM transformer
+ * @param {Object} markupconfig the markup config
+ */
+async function adjustMDAST(context, { logger, transformer, markupconfig }) {
+  if (!markupconfig || !markupconfig.markup) {
+    return;
+  }
+
+  Object.entries(markupconfig.markup)
+    .filter((entry) => entry[1].type === 'markdown')
+    .forEach((entry) => {
+      logger.info(`Applying markdown markup adjustment: ${entry[0]}`);
+
+      const cfg = entry[1];
+      transformer.match(cfg.match, (h, node) => {
+        const handler = transformer.constructor.default(node);
+        let el = handler(h, node);
+
+        // Append classes to the element (space or comma separated)
+        if (cfg.classnames) {
+          el.properties.className = [
+            ...(el.properties.className || '').split(' '),
+            ...cfg.classnames,
+          ].join(' ').trim();
+        }
+
+        // Append attributes to the element
+        if (cfg.attribute) {
+          Object.assign(el.properties, cfg.attribute);
+        }
+
+        // Wrap the element
+        if (cfg.wrap) {
+          const wrapperEl = getHTMLElement(cfg.wrap);
+          const n = fromDOM(wrapperEl);
+          el = findAndReplace(n, PLACEHOLDER_TEMPLATE, () => el);
+        }
+
+        return el;
+      });
+    });
+}
+
+/**
+ * Adjust the DOM tree according to the markup config.
+ * This is done by directly manipulating the DOM after it has been generated by the pipeline.
+ *
+ * @param {Object} context the execution context
+ * @param {Object} logger the pipeline logger
+ * @param {Object} markupconfig the markup config
+ */
+async function adjustHTML(context, { logger, markupconfig }) {
+  if (!markupconfig || !markupconfig.markup) {
+    return;
+  }
+
+  Object.entries(markupconfig.markup)
```

Comment:
Since https://github.com/adobe/helix-shared/releases/tag/v5.0.1 you should be able to treat `markupconfig.markup` as an array, which should make it a bit easier.

Reply:
It's still coming in as an object, which is fine since this gives me access to the keys to help debugging

Reply:
`entry.name` should also contain the name in case you need it for debugging. Reading it as an array might also help in keeping the order stable, which isn't guaranteed for objects.

Reply:
Doesn't validate at the moment as `.name` is not part of the schema. I filed adobe/helix-shared#248 to track this